File: exercises/polimorfism/Figura.cpp
#include "Figura.h"

Figura::Figura(std::string nome, float dim1)
    : nome(nome), dim1(dim1) {}
Figura::Figura(std::string nome, float dim1, float dim2)
    : nome(nome), dim1(dim1), dim2(dim2) {}
Figura::Figura(std::string nome, float dim1, float dim2, float dim3)
    : nome(nome), dim1(dim1), dim2(dim2), dim3(dim3) {}

Figura::~Figura() {}

std::string Figura::getNome() const {return nome;}
float Figura::getDim1() const {return dim1;}
float Figura::getDim2() const {return dim2;}
float Figura::getDim3() const {return dim3;}

void Figura::setNome(const std::string& novoNome) {nome = novoNome;}
void Figura::setDim1(float novoDim1) {dim1 = novoDim1;}
void Figura::setDim2(float novoDim2) {dim2 = novoDim2;}
void Figura::setDim3(float novoDim3) {dim3 = novoDim3;}

void Figura::aumentar(float percentage) {
    // o % na questão interpretei como se fosse de 0 a 100
    float adjusted = percentage/100.0f; // qualquer coisa só comentar essa linha

    setDim1(getDim1()*adjusted);
    setDim2(getDim2()*adjusted);
    setDim3(getDim3()*adjusted);
}

File: java/polimorfism_java/Figura.java
public abstract class Figura {
    protected String nome;
    protected float dim1;
    protected float dim2;
    protected float dim3;

    public Figura(String nome, float dim1) {
        this.nome = nome;
        this.dim1 = dim1;
        this.dim2 = 0;
        this.dim3 = 0;
    }

    // Construtor para Figura com dim1 e dim2
    public Figura(String nome, float dim1, float dim2) {
        this.nome = nome;
        this.dim1 = dim1;
        this.dim2 = dim2;
        this.dim3 = 0;
    }

    // Construtor para Figura com dim1, dim2 e dim3
    public Figura(String nome, float dim1, float dim2, float dim3) {
        this.nome = nome;
        this.dim1 = dim1;
        this.dim2 = dim2;
        this.dim3 = dim3;
    }

    public abstract void display();
    public abstract float getArea();
    public abstract float getVolume();

    // Getters e Setters
    public String getNome() {return nome;}

    public float getDim1() {return dim1;}

    public float getDim2() {return dim2;}

    public float getDim3() {return dim3;}

    public void setNome(String novoNome) {this.nome = novoNome;}

    public void setDim1(float novoDim1) {this.dim1 = novoDim1;}

    public void setDim2(float novoDim2) {this.dim2 = novoDim2;}

    public void setDim3(float novoDim3) {this.dim3 = novoDim3;}

    // Método para aumentar as dimensões em uma porcentagem
    public void aumentar(float percentage) {
        this.dim1 *= (1+percentage/100);
        this.dim2 *= (1+percentage/100);
        this.dim3 *= (1+percentage/100);
    }
}